### ISSUE TITLE
Electric board channels: multiple values in an array

### ISSUE THREAD

**Issue by xen2** [scottyphillips/echonetlite_homeassistant#117]
Great work on the project!

I was working on adding my Panasonic "Smart Cosmo" electric distribution board to Home Assistant.

The thing is, one single ECHONET value returns up to 60 values in an array.

On the pychonet side, I wrapped the 60 values in a JSON array.

On the home assistant plugin side, I did the following "hack":
https://github.com/scottyphillips/echonetlite_homeassistant/commit/56c3c93993d4143caa524dffe584be7152430fdd#diff-e82565bd3ee78c190e10f8e620df9a97630dfb81dc19fd88f2aeefedc785586dR86

Basically, for a single ECHONET value, I allow the creation of multiple sensors with a "subpath" (resolved by a lambda) on the ECHONET property.
Does it seem like a good approach, or do you have any other idea?
Note: if we stick with this lambda approach, it could be used in various other cases to simplify a bunch of if/else for special cases, unit conversion, etc.

Anyway, using this approach everything is working, I can see the consumption of each of my individual 40 lines in watts.
![image](https://user-images.githubusercontent.com/527565/193462385-5a85188c-e83e-40de-8072-7581fdb8bc77.png)

Happy to do a PR once everything is working and clean!

**Comment by scottyphillips** [comment]
Please do so!

**Comment by nao-pon** [comment]
@xen2 Thank you for your contribution. it looks good. One question is why did you include the lambda value instead of just the index. Is there anything wrong with getting by index from a dictionary object?

Also, I would like to know the difference of pychonet.

Thanks! 👍

**Comment by xen2** [comment]
Thanks for the follow-up!
The reason I used a lambda is:
* For this specific case, I needed to access `value['values'][index]` but the JSON message could have different names/structures
* It could also be useful for type conversion (i.e. some kW to W) in the future.
  For example, some of the special cases hardcoded in `EchonetSensor.native_value` could be directly handled with this system.

I have published the pychonet branch:
https://github.com/xen2/pychonet/tree/distribpanel_individual_channels

I will be doing a bunch of other patches soon, I have a bunch of other devices to connect (enefarm, solar panel, batteries, etc.).

**Comment by nao-pon** [comment]
I checked the specs and it seems that out of the number of available channels (252), the maximum is 60 channels starting from a specific channel number. Need information on this channel number?

In my implementation, the starting channel number and the number appended to the entity name displayed in the HA may not match.

Regarding the specification of the unit, the value specified by CONF_UNIT_OF_MEASUREMENT has high priority.
- https://github.com/scottyphillips/echonetlite_homeassistant/commit/9ef2ad1aeeaffdfb294ee912439a6148a0ee60a2

**Comment by xen2** [comment]
In my case I needed only 40 so I didn't bother to handle "pagination".

Otherwise, what needs to be done is splitting in multiple requests:
Query number of channels with 0xB1
![image](https://user-images.githubusercontent.com/527565/193951395-637c81fa-5c8f-4d2f-902b-18fe106850b9.png)

Then start a loop (assuming 100 values, it would do ranges 0-39, 40-79, 80-99)
- set desired range with 0xB6
- then query values with 0xB7
![image](https://user-images.githubusercontent.com/527565/193951214-863ce583-5939-42bc-91ea-5759fe223908.png)

Even if we don't implement all of that, it would be good to at least do `+ value.start` (which should be 0 only so far) so that it doesn't break if we add pagination later.

**Comment by xen2** [comment]
English manual for reference:

![image](https://user-images.githubusercontent.com/527565/193951718-c7f9a855-7d05-4fec-85df-5267dba57372.png)
![image](https://user-images.githubusercontent.com/527565/193951741-793bc77d-b6e2-4594-9cfa-c8563e86051a.png)

Sources:
EN: https://echonet.jp/wp/wp-content/uploads/pdf/General/Standard/Release/Release_Q/Appendix_Release_Q_E.pdf
JP: https://echonet.jp/wp/wp-content/uploads/pdf/General/Standard/Release/Release_Q/Appendix_Release_Q.pdf

**Comment by xen2** [comment]
> Does the value of EPC:0xB1 match the actual number of circuit breakers you are using? If so, creating that number of entities at startup eliminates wasted entities.

Yes. Or even easier, it's part of the result of 0xB7:
`{ start: 0, range: 40, values: { .... 40 values here ... } }`

Assuming you don't plan to support values past 60, you can use the `range: 40` result part of 0xB7 answer.
That's why I did `index: value['values'][index] if index < value['range'] else None` in my lambda: it returns None if values are out of bound, so even though 60 sensors are created (because I don't know at creation time), the last 20 are disabled.
Of course if we could ask 0xB1 and wait answer **before** sensor are created, that would be even better (no need to create non-existing sensors 41..60).

**Comment by nao-pon** [comment]
I'm considering a config like this:

```python
        0x87 : {
            0xB7: {
                CONF_TYPE: DEVICE_CLASS_POWER,
                CONF_STATE_CLASS: SensorStateClass.MEASUREMENT,
                TYPE_DATA_DICT: {
                    "getnum": 0xB1,
                    "setdata": {
                        0xB6: 0
                    }
                }
            },
```

"getnum" is the EPC number to get the maximum value when the returned dictionary is an arbitrary number.

"setdata" will be checked before GET and sent in advance if different.

I'm still thinking about the implementation. Please let me know if you have any suggestions.

**Comment by xen2** [comment]
Looks good for the config. Let's see how we can generalize it to other cases later.
What does the `0xB6: 0` mean? is this the start index to set? I would name it `setrange`. In practice I don't need to set it, it returns 0..40 by default in my case.
Anyway, we probably can remove the 0xB6 setrange for now (cf next part). 0xB1 should be enough if we don't support past 60.

Ideally, I would say that we would like to run the 0xB1 command during the device discovery phase (`async_setup_entry()`). It looks like we could use `response = await self._api.echonetMessage(...)` inside `if TYPE_ARRAY_MAX_SIZE in _keys:` block of `async_setup_entry()` (which is async) to achieve that.
I don't think we need to call 0xB6 at all, we can just log a warning if 0xB1 is larger than 60, saying we won't be creating/recording sensors past the 60th.

Then we can proceed with the sensor creation loop once we know the exact number.

**Comment by nao-pon** [comment]
Oh, 0xB6 should also include the range to get. I am concerned that other devices may change the value of 0xB6. You can limit the range to 60. For this library, just throw it away.

It's `0xB6: b'\x00\x3C'`.

**Comment by xen2** [comment]
Good point, it could be changed behind our back.

I don't think you can set the range to 60 if only 40 channels, you probably have to set a range that is max 0..40, not 0..60

I am not sure if possible, but to avoid another setrange being set behind our back, maybe there is a way to do a request that set the 0xB6 value and read 0xB7 in a single atomic transaction? I took a quick look at the spec and found a `SetGet` but that's probably only for the same property.
Otherwise, simply setting 0xB6 once and reading 0xB7 is probably good enough.

Either way, even if we don't call 0xB6 setrange but only 0xB7, we can still be safe if we interpret both "start" and "range" from 0xB7 and use it when updating values, i.e. lambda could be:
`value['values'][index - value['start']] if index >= value['start'] && index < value['start'] + value['range'] else None`
That way, we can at least be sure to not update the wrong index if "start" is not 0 (worst thing that can happen is that we don't update 0..60 anymore, but since we don't have case with 60+ channels yet, this shouldn't happen yet -- we can even issue warnings that it's not supported yet if it happens).

**Comment by nao-pon** [comment]
@xen2 After much consideration, I think your implementation is currently the simplest.

For the time being, I would like to support such properties as long as they are not paginated, and if there are any more, I would like to be undefined.

I look forward to your PR. thank you! 👍

**Comment by nao-pon** [comment]
Earlier, I merged the implementation of TYPE_DATA_DICT, so I think that it corresponds to the values of EPC 0xC6, 0xC7 of 0x02-0x87.

Could you please check it as well?

const.py
```python
            0xC7: {
                CONF_TYPE: DEVICE_CLASS_CURRENT,
                CONF_STATE_CLASS: SensorStateClass.MEASUREMENT,
                TYPE_DATA_DICT: ["r_phase_amperes", "t_phase_amperes"]
            },
            0xC8: {
                CONF_TYPE: DEVICE_CLASS_VOLTAGE,
                CONF_STATE_CLASS: SensorStateClass.MEASUREMENT,
                TYPE_DATA_DICT: ["r_sn_voltage", "sn_t_voltage"]
            },
```

**Comment by xen2** [comment]
Note: I didn't forget about this PR but waiting for the second electric board to be setup to finish it.

**Comment by nao-pon** [comment]
@xen2 I merged your branch electric_board_channels into master and created a PR with enhancements for the "light", "fan" entities.

You will need to use my edge branch to test this as it also requires an update to pychonet for it to work. Could you please verify using it?

**Comment by xen2** [comment]
Thanks! I will give it a try later (hopefully next week).

**Comment by nao-pon** [comment]
@xen2 I have a question. Only WTY2001 gets uidi from 0xFE, what is the reason?
Is the reason for not using f"{uid}-{eojgc}-{eojcc}-{instance}" a change in the instance number of the device managed by this device?

https://github.com/scottyphillips/echonetlite_homeassistant/blob/0ac92848df1bdd6c238cd23187ca3cc6edbe8874/custom_components/echonetlite/config_flow.py#L189-L205

**Comment by rubyroobs** [comment]
> @xen2 I have a question. Only WTY2001 gets uidi from 0xFE, what is the reason? Is the reason for not using f"{uid}-{eojgc}-{eojcc}-{instance}" a change in the instance number of the device managed by this device?
> 
> https://github.com/scottyphillips/echonetlite_homeassistant/blob/0ac92848df1bdd6c238cd23187ca3cc6edbe8874/custom_components/echonetlite/config_flow.py#L189-L205

Hi @xen2 - I wanted to bump @nao-pon's question here. I use the same lights and this change required me to reconfigure all my lights after updating to use this to get the `uidi`. It is neat that the name comes directly from what you set in the Link Plus app, but is there a particular reason the `uidi` itself needs to be different? It would be nice to confirm so there can be consistent behavior over updates :D

**Comment by rubyroobs** [comment]
> Sorry for late answer and that it broke your config.

@xen2 no problem at all! Thank you for the very detailed explanation and contributing the change in the first place. I agree this makes sense to switch over to the hardware based IDs for the switches in the way you configured. @nao-pon if you're ok with this too, can we change the `edge` branch back to using this?

Separately, thanks for sharing your work and sorry to hear you have had the experience of using their switch app also 🤣 For me, as soon as I got them added (which was enough of a pain by itself with their convoluted "工事メニュー" and holding the button every 5 minutes etc) I just controlled everything in HA and made groups in there. I'd honestly be interested in outright replacing the WTY2001 with something running ESPHome or similar but AFAIK there is no work done to reverse engineer their protocol (and bluetooth is not ideal for latency). It's a shame that Lutron etc do not sell in Japan - I had to run so many neutral wires installing my Link Plus switches anyway I would have rather just done that for all of them and used something better...
